docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+
+The desire for sandboxing also leads to the conclusion that a 'foreign function interface' is exposing the wrong level of abstraction.
+What sandbox should a function run in?
+Our prior [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work showed that treating libraries (in the loose sense, including any dependent libraries) as the unit of foreign code provided an abstraction that fitted well with sandboxing and allowed the user to determine the degree of state sharing between foreign code invocations.
+This abstraction also has software engineering benefits.
+LLVM, for example, does not have stable ABI (or API) guarantees across programs and so can cause linking errors if, for example, an application depends on one library that uses libclang from LLVM 10 to parse syntax highlighting and another that uses the JIT from LLVM 11 for GPU shaders.
+In a Verona world, these would be in different sandboxes with different 'global' namespaces and so would not conflict.
+
+### Sandboxes are surfaced as regions
+
+Sandboxing is a security *policy* that can be implemented with a variety of *mechanisms*.
+Sandboxing runs a component with a reduced set of privileges and limits its interactions with the outside world.
+It can be enforced with a separate process that runs with reduced OS privilege (e.g. the OpenSSH privesp model or Chromium's renderer sandboxes), with SFI techniques such as those used in WebAssembly compilers or NaCl (potentially with hardware acceleration from features such as Intel MPK), or with in-address-space hardware isolation such as CHERI.
+
+Sandboxes are difficult to program with in C-like languages because there is no useful construct in the abstract machine that can represent them.
+Web browsers are the most aggressively compartmentalised applications and yet run with a handful of different sandbox types.
+Because there is no abstraction in C for distinct sets of objects with mediated entry points, sandboxing code written in C leaks mechanism into the security code (for example, the idea that sandboxes are process).
+Any call between sandboxes must go via some RPC layer.
+Allocating memory inside a sandbox from outside and directly manipulating data structures is difficult to get right because the language doesn't impose any restrictions on where any given pointer can be stored.
+
+A Verona region is a set of related objects, with arbitrary topology, that have a single entry point.
+Sandboxes map cleanly to this abstraction.
+A sandbox's heap can contain pointers to any other data structures owned by the sandbox.
+It cannot contain pointers outside of the sandbox because following them would violate the sandbox policy.
+The Verona type system enforces this automatically by disallowing cross-region pointers.
+
+We hope that the region abstraction will make programming with sandboxes natural within the language.
```

Comment:
Creating a region and loading a shared library.  The region may or may not be wrapped in a single-region cown.

Reply:
Note: It is fine for Verona *code* to operate on data in a foreign region, as long as the compiler inserts the correct dynamic checks.  It is not fine to depend on type safety for Verona *objects* in a foreign region.  We will have to impose some constraints on the kinds of object that can be allocated in a foreign region.

Reply:
"We will have to impose some constraints on the kinds of object that can be allocated in a foreign region."

Can you elaborate on that more?

Reply:
I thought all Verona objects would be allocated by the stub caller (for arguments and return values) and not the external code. I believe that's what JNI used to do. I don't know how we could guarantee that the external code hasn't messed up a Verona object internally if it's in its own memory.

Reply:
See the process isolation demo.  This allows allocating objects in the child from the parent *without* an RPC.  We don't aim to guarantee that the object hasn't been messed up, we aim to guarantee that messing up the object only lets you corrupt other objects within the sandbox.

Reply:
@SeanTAllen, any object that is allocated in a sandbox region becomes untrusted immediately after any untrusted code is allowed to execute in the untrusted region (which, with process sandboxing, is immediately, unless we have OS support for suspending all threads in a child process - this is much easier with SFI or CHERI sandboxes, where we control their scheduling).  This means that we need to check any pointers that we follow and we can't trust that they haven't been tampered with.  For example, we can't trust a vtable pointer for a verona object.  We may be able to load the vtable pointer, validate it, and then invoke a method, but we'll need the compiler to generate methods that check access.

We especially need to be careful of tearing and TOCTOU vulnerabilities, because malicious code in the sandbox can potentially race you in accessing different fields in an object, so any pair of fields that need accessing atomically will need to be copied out and the copies used.  

We probably need to start with some *very* tight constraints on what kinds of objects can be allocated in untrusted regions.  Something similar to C++'s POD types: no dynamic dispatch, all fields are primitive or point to other objects with the same restrictions.